src/core/markovProcessGraph.js

```diff
@@ -0,0 +1,605 @@
+  // next epoch node for the same owner.
+  +gammaForward: TransitionProbability,
+  +gammaBackward: TransitionProbability,
+|};
+export type SeedOptions = {|
+  +alpha: TransitionProbability,
+|};
+
+const COMPAT_INFO = {type: "sourcecred/markovProcessGraph", version: "0.1.0"};
+
+export type MarkovProcessGraphJSON = Compatible<{|
+  +nodes: {|+[NodeAddressT]: MarkovNode|},
+  +edges: {|+[MarkovEdgeAddressT]: MarkovEdge|},
+  +scoringAddresses: $ReadOnlyArray<NodeAddressT>,
+|}>;
+
+export class MarkovProcessGraph {
+  _nodes: Map<NodeAddressT, MarkovNode>;
+  _edges: Map<MarkovEdgeAddressT, MarkovEdge>;
+  _scoringAddresses: Set<NodeAddressT>;
+
+  constructor(
+    nodes: Map<NodeAddressT, MarkovNode>,
+    edges: Map<MarkovEdgeAddressT, MarkovEdge>,
+    scoringAddresses: Set<NodeAddressT>
+  ) {
+    this._nodes = nodes;
+    this._edges = edges;
+    this._scoringAddresses = scoringAddresses;
+  }
+
+  static new(
+    wg: WeightedGraphT,
+    fibration: FibrationOptions,
+    seed: SeedOptions
+  ) {
+    const _nodes = new Map();
+    const _edges = new Map();
+    const _scoringAddresses = _findScoringAddresses(wg.graph, fibration.what);
+
+    // _nodeOutMasses[a] = sum(e.pr for e in edges if e.src == a)
+    // Used for computing remainder-to-seed edges.
+    const _nodeOutMasses = new Map();
+
+    const epochTransitionRemainder = (() => {
+      const {alpha} = seed;
+      const {beta, gammaForward, gammaBackward} = fibration;
+      if (beta < 0 || gammaForward < 0 || gammaBackward < 0) {
+        throw new Error(
+          "Negative transition probability: " +
+            [beta, gammaForward, gammaBackward].join(" or ")
+        );
+      }
+      const result = 1 - (alpha + beta + gammaForward + gammaBackward);
+      if (result < 0) {
+        throw new Error("Overlarge transition probability: " + (1 - result));
+      }
+      return result;
+    })();
+
+    const timeBoundaries = (() => {
+      const edgeTimestamps = Array.from(
+        wg.graph.edges({showDangling: false})
+      ).map((x) => x.timestampMs);
+      const start = min(edgeTimestamps);
+      const end = max(edgeTimestamps);
+      const boundaries = weekIntervals(start, end).map((x) => x.startTimeMs);
+      return [-Infinity, ...boundaries, Infinity];
+    })();
+
+    const addNode = (node: MarkovNode) => {
+      if (_nodes.has(node.address)) {
+        throw new Error("Node conflict: " + node.address);
+      }
+      _nodes.set(node.address, node);
+    };
+    const addEdge = (edge: MarkovEdge) => {
+      const mae = rawEdgeAddress(edge);
+      if (_edges.has(mae)) {
+        throw new Error("Edge conflict: " + mae);
+      }
+      const pr = edge.transitionProbability;
+      if (pr < 0 || pr > 1) {
+        const name = MarkovEdgeAddress.toString(mae);
+        throw new Error(`Invalid transition probability for ${name}: ${pr}`);
+      }
+      _edges.set(mae, edge);
+      _nodeOutMasses.set(edge.src, _nodeOutMasses.get(edge.src) || 0 + pr);
```

Comment:
Umm, I believe that addition is actually higher precedence than ||. so i think this code is wrong.
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Operator_Precedence#:~:text=Operator%20precedence%20determines%20how%20operators,of%20operators%20with%20lower%20precedence.